kratos/mpi/includes/mpi_message.h

```diff
@@ -86,113 +73,409 @@ template<> struct MPIDataType<char>
+
+    using SubDataType = TDataType;
+
+    using PrimitiveDataType = typename MPIMessage<TDataType>::PrimitiveDataType;
+
+    static constexpr bool HasContiguousPrimitiveData = std::is_same_v<TDataType, PrimitiveDataType>;
+
+    static constexpr bool HasDynamicMemoryAllocation = true;
+
+    inline MPI_Datatype DataType()
+    {
+        return Internals::MPIDataType<PrimitiveDataType>::DataType();
+    }
+
+    inline void* Buffer(MessageDataType& rValue)
+    {
+        if constexpr(HasContiguousPrimitiveData) {
+            return rValue.data();
+        } else {
+            if (rValue.size() > 0) {
+                MPIMessage<TDataType> sub_mpi_message;
+                const auto sub_data_type_size = sub_mpi_message.Size(rValue.front());
+
+                const auto primitive_data_size = Size(rValue);
+                if (mTempValues.size() != static_cast<unsigned int>(primitive_data_size)) {
+                    mTempValues.resize(primitive_data_size);
+                }
+
+                unsigned int local_index = 0;
+                for (unsigned int i = 0; i < rValue.size(); ++i) {
+                    PrimitiveDataType* begin = static_cast<PrimitiveDataType*>(sub_mpi_message.Buffer(rValue[i]));
+                    for (int j = 0; j < sub_data_type_size; ++j) {
+                        mTempValues[local_index++] = *(begin++);
+                    }
+                }
+            } else {
+                mTempValues.resize(0);
+            }
+
+            return mTempValues.data();
+        }
+    }
+
+    inline const void* Buffer(const MessageDataType& rValue)
+    {
+        if constexpr(HasContiguousPrimitiveData) {
+            return rValue.data();
+        } else {
+            if (rValue.size() > 0) {
+                MPIMessage<TDataType> sub_mpi_message;
+                const auto sub_data_type_size = sub_mpi_message.Size(rValue.front());
+
+                const auto primitive_data_size = Size(rValue);
+                if (mTempValues.size() != static_cast<unsigned int>(primitive_data_size)) {
+                    mTempValues.resize(primitive_data_size);
+                }
+
+                unsigned int local_index = 0;
+                for (unsigned int i = 0; i < rValue.size(); ++i) {
+                    PrimitiveDataType const* begin = static_cast<PrimitiveDataType const*>(sub_mpi_message.Buffer(rValue[i]));
+                    for (int j = 0; j < sub_data_type_size; ++j) {
+                        mTempValues[local_index++] = *(begin++);
+                    }
+                }
+            } else {
+                mTempValues.resize(0);
+            }
+
+            return mTempValues.data();
+        }
+    }
+
+    inline int Size(const MessageDataType& rValue)
+    {
+        return rValue.size() * SubDataTypeSize(rValue);
+    }
+
+    inline int SubDataTypeSize(const MessageDataType& rValue)
+    {
+        return rValue.size() == 0 ? 0 : MPIMessage<SubDataType>().Size(rValue.front());
```

Comment:
@jcotela I would greately appreciate if you can review this, I can put those remaks on a seperate PR. :)

Reply:
@sunethwarna other than what I said above, I only have a couple of comments regarding MPIMessage
- there is a `template<class TDataType> class MPIMessage;` in line 88 of mpi_message.h that is not needed
- Why do you need to instance the MPIMessage class? I would have kept its methods static, since I do not see it manage any data at the moment.

I am quite happy with how this turned out overall, since you managed to fit the new features within the existing framework.

Reply:
@jcotela Thanks a lot for the comments. :D. Again, nice to have you back in Kratos :D

> there is a template<class TDataType> class MPIMessage; in line 88 of mpi_message.h that is not needed

Indeed, I missed it. I will fix it :)

> Why do you need to instance the MPIMessage class? I would have kept its methods static, since I do not see it manage any data at the moment.

The `MPIMessage` cannot be static in the case of non-contiguous memory arrays such as `std::vector<array_1d<double, 3>>` or `std::vector<Vector>` which is now supported. I was careful enough to reduce memeory and computational overheads while supporting these non-contiguous vectors. In there, `MPIMessage::Buffer` returns the contiguous memeory for MPI communication (if it is contiguous, then it does not do anything, returns the original data array. If it is not contiguous, then it creates a contiguous array from the original data and sends that memory location.). So because we now support non-contiguous arrys, there is this `MPIMessage::Update` method, which updates the original non-contiguous arrays given by user with the MPI communicated data.

This is how it looks like in a simple `AllReduceDetail`
https://github.com/KratosMultiphysics/Kratos/blob/972ae53b9f5984bde778088986d45f682eccbc5b/kratos/mpi/sources/mpi_data_communicator.cpp#L635-L659

Data management is done in places like following (Buffer):
https://github.com/KratosMultiphysics/Kratos/blob/972ae53b9f5984bde778088986d45f682eccbc5b/kratos/mpi/includes/mpi_message.h#L367-L394

Update method:
https://github.com/KratosMultiphysics/Kratos/blob/972ae53b9f5984bde778088986d45f682eccbc5b/kratos/mpi/includes/mpi_message.h#L452-L476

Reply:
Ok I see why it is needed. I guess my first preference in this case would have been to implement the buffering and updating explicitly on the caller function, but I can see how this helps and may reduce code duplication.

Looking into #11440  now, I think there is something going on with single-value sends of type Matrix (not std::vector<Matrix>). These are going through the serializer in the current master, but if I try to define a Send for Matrix using the mpi_data_communicator type handling the TestMPIDataCommunicatorSendAndRecvMatrix suddenly fails.